### PR TITLE
Workflow to deploy merged commit to production

### PR DESCRIPTION
**Pull request by vigneshmsft**
### Changes proposed in this pull request

build.yml builds docker images and triggers a deployment workflow(deploy.yml)

Deployment workflow starts a GitHub deployment to QA,
triggers smoke test on the deployed environment and then triggers deployment to the successive environment.

QA -> Staging -> Production

deploy.yml also has a manual workflow dispatch trigger to enable manual deployments(for rollback etc)

### Trello card
https://trello.com/c/A4528lno/2382-deploy-to-staging-and-production-in-paas
### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally
- [x] Product review

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -54,9 +54,9 @@ jobs:
...
         if: ${{ success() }}
-        run: |
-          curl -X POST \
-               -H "Accept: application/vnd.github.v3+json" \
-               -H "Authorization: Bearer ${{ secrets.ACTIONS_API_ACCESS_TOKEN }}" \
-               https://api.github.com/repos/DFE-Digital/find-teacher-training/deployments \
-               -d '{"ref":"${{ github.sha }}", "environment": "qa", "required_contexts": []}'
+        uses: benc-uk/workflow-dispatch@v1.1
+        with:
+          workflow: 3074188 # deploy.yml from https://api.github.com/repos/DFE-Digital/find-teacher-training/actions/workflows
```
Comment (saliceti) [inline .github/workflows/build.yml:59]: Can we use the workflow name instead? As per https://github.com/benc-uk/workflow-dispatch#workflow

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -54,9 +54,9 @@ jobs:
...
+          token: ${{ secrets.ACTIONS_API_ACCESS_TOKEN }}
+          inputs: '{"environment": "qa", "sha": "${{ github.sha }}"}'
+          ref: ${{ github.ref }}
```
Comment (saliceti) [inline .github/workflows/build.yml:62]: Do we need to add the ref? Looks like it uses the context ref by default: https://github.com/benc-uk/workflow-dispatch#ref

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -1,24 +1,50 @@
 name: Deploy to PaaS
 
 on:
-  deployment
+  workflow_dispatch:
+    inputs:
+      environment:
+        description: 'The environment to deploy to (qa, staging or production)'
```
Comment (saliceti) [inline .github/workflows/deploy.yml:7]: We might use different environment names in the future. Maybe:
```
The environment to deploy to, eg: qa, staging, production...
```
Also please remove the quotes

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -46,28 +72,39 @@ jobs:
...
-          curl -X POST \
-            https://api.github.com/repos/DFE-Digital/find-teacher-training-tests/actions/workflows/smoke-tests.yml/dispatches \
-            -H "Accept: application/vnd.github.v3+json" \
-            -H "Authorization: Bearer ${{ secrets.ACTIONS_API_ACCESS_TOKEN }}" \
-            -d '{"ref": "refs/heads/master", "inputs": {"environment": "${{ env.DEPLOY_ENV }}"} }'
+      - name: Trigger ${{ env.NEXT_ENV }} Deployment
+        if: ${{ steps.wait_for_smoke_tests.outputs.conclusion == 'success' && env.DEPLOY_ENV != 'production' }}
+        uses: benc-uk/workflow-dispatch@v1.1
+        with:
+          workflow: 3074188 # deploy.yml from https://api.github.com/repos/DFE-Digital/find-teacher-training/actions/workflows
```
Comment (saliceti) [inline .github/workflows/deploy.yml:97]: Can we use the workflow name instead?

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -46,28 +72,39 @@ jobs:
...
           TF_VAR_paas_app_docker_image: ${{ env.DOCKER_IMAGE }}
 
-      - name: Post Deployment Status
-        run: |
-          current_deployment=$(curl -s -X GET -H "Authorization: Bearer ${{ github.token }}" 'https://api.github.com/repos/DFE-Digital/find-teacher-training/deployments?ref=${{ github.event.deployment.ref }}&environment=${{ github.event.deployment.environment }}')
-          deployment_id=$(echo $current_deployment | jq '.[0].id')
+      - name: Trigger ${{ env.DEPLOY_ENV }} Smoke Tests
+        uses: benc-uk/workflow-dispatch@v1.1
+        with:
+          workflow: 3153670 # smoke-test.yml from https://api.github.com/repos/DFE-Digital/find-teacher-training/actions/workflows
```
Comment (saliceti) [inline .github/workflows/deploy.yml:78]: Can we use the workflow name instead?

File: .github/workflows/build.yml (hunk 1 above)
Comment (vigneshmsft) [reply .github/workflows/build.yml]: done, but we have to remember to update references if the name changes.

**Review by saliceti — APPROVED** [review]

**Pull request merged**